internal/databases/mongo/binary/storage.go

```diff
@@ -0,0 +1,162 @@
+	return internal.UploadSentinel(backupStorage.UploaderProvider, sentinel, backupStorage.BackupName)
+}
+
+func (backupStorage *BackupStorage) UploadMongodBackupFilesMetadata(filesMetadata *MongodBackupFilesMetadata) error {
+	filesMetadataPath := backupStorage.FilesMetadataNameFromBackup()
+	return internal.UploadDto(backupStorage.UploaderProvider.Folder(), filesMetadata, filesMetadataPath)
+}
+
+func (backupStorage *BackupStorage) DownloadSentinel() (*models.Backup, error) {
+	return common.DownloadSentinel(backupStorage.UploaderProvider.Folder(), backupStorage.BackupName)
+}
+
+func (backupStorage *BackupStorage) DownloadMongodBackupFilesMetadata() (*MongodBackupFilesMetadata, error) {
+	var filesMetadata MongodBackupFilesMetadata
+	filesMetadataPath := backupStorage.FilesMetadataNameFromBackup()
+	err := internal.FetchDto(backupStorage.UploaderProvider.Folder(), &filesMetadata, filesMetadataPath)
+	if err != nil {
+		return nil, errors.Wrap(err, "can not fetch files metadata")
+	}
+	return &filesMetadata, nil
+}
+
+func (backupStorage *BackupStorage) FilesMetadataNameFromBackup() string {
+	return filepath.Join(backupStorage.BackupName, FilesMetadataFileName)
+}
+
+func (backupStorage *BackupStorage) MakeBackupFilePath(backupFileMeta *BackupFileMeta) string {
+	return filepath.Join(backupStorage.BackupName, backupFileMeta.Path+"."+backupFileMeta.Compression)
+}
+
+func (backupStorage *BackupStorage) CreateReader(backupFileMeta *BackupFileMeta) (io.ReadCloser, error) {
+	objectPath := backupStorage.MakeBackupFilePath(backupFileMeta)
+	objectReader, err := backupStorage.UploaderProvider.Folder().ReadObject(objectPath)
+	if err != nil {
+		return objectReader, err
+	}
+	decompressor := compression.FindDecompressor(backupFileMeta.Compression)
+	if decompressor == nil {
+		return nil, fmt.Errorf("decompressor for %v not found", backupFileMeta.Compression)
+	}
+	return internal.DecompressDecryptBytes(objectReader, decompressor)
+}
+
+func (backupStorage *BackupStorage) GetCompression() string {
+	return backupStorage.UploaderProvider.Compression().FileExtension()
+}
+
+// todo: need multitraded upload
+func (backupStorage *BackupStorage) UploadFile(reader io.Reader, backupFileMeta *BackupFileMeta) error {
+	destinationFilePath := backupStorage.MakeBackupFilePath(backupFileMeta)
+
+	return backupStorage.UploaderProvider.PushStreamToDestination(reader, destinationFilePath)
+}
+
+func (backupStorage *BackupStorage) CalculateCompressedFiles(backupFiles map[string]*BackupFileMeta) (int64, error) {
+	folder := backupStorage.UploaderProvider.Folder().GetSubFolder(backupStorage.BackupName)
+	return backupStorage.calculateCompressedFiles("", folder, backupFiles)
+}
+
+//nolint: whitespace
+func (backupStorage *BackupStorage) calculateCompressedFiles(relativePath string, folder storage.Folder,
+	backupFiles map[string]*BackupFileMeta) (backupSize int64, err error) {
+
+	objects, subFolders, err := folder.ListFolder()
```

Comment:
Why list folder to calculate the size? We can take this information from the appropriate `Uploader`

Reply:
I implement like internal.FolderSize, that is used in logical backup. 
Besides I set compressed size for each file to save in meta (for troubleshooting purpose).
I of cause can remove it and use uploader.UploadedDataSize(), but I will be forced to remove compressed size for each my file.

Reply:
Since we upload files one by one in the current version, these S3 listings might be massive.